src/python/pants/base/build_root.py

```diff
@@ -51,25 +52,23 @@ def path(self):
     return self._root_dir
 
   @path.setter
-  def path(self, root_dir):
+  def path(self, root_dir: str) -> None:
     """Manually establishes the build root for the current workspace."""
     path = os.path.realpath(root_dir)
     if not os.path.exists(path):
       raise ValueError(f'Build root does not exist: {root_dir}')
     self._root_dir = path
 
-  def reset(self):
+  def reset(self) -> None:
     """Clears the last calculated build root for the current workspace."""
     self._root_dir = None
 
-  def __str__(self):
+  def __str__(self) -> str:
     return f'BuildRoot({self._root_dir})'
 
   @contextmanager
-  def temporary(self, path):
+  def temporary(self, path: str) -> Iterator[None]:
     """Establishes a temporary build root, restoring the prior build root on exit."""
-    if path is None:
```

Comment:
Does mypy type checking actually have the flow analysis to safely remove this check? I have a hard time believing it can. This has a mypy committee admitting it's not perfect fairly recently: https://github.com/python/mypy/issues/6477

I'd steer away from deleting runtime checks unless you have some greater context showing they're not needed above and beyond mypy maybe making them un-needed under most circumstances.

Reply:
I agree, good point. We especially shouldn't remove runtime checks until we have full MyPy coverage, which we are still far from.